File: solutions/cpp/run-length-encoding/1/run_length_encoding.cpp
#include "run_length_encoding.h"

using namespace std;

namespace run_length_encoding {
    

string encode(string text){
    string res;
    int count=0;
    char p=0;
    for(char c : text){
       if(p==c) {
            count++;
        }        
        else {
            if(p)
                res += count<=1 ? string{p} : to_string(count) + p;
            count = 1;
            p = c;           
        }
    }    
    if(p)
        res += count<=1 ? string{p} : to_string(count) + p;
    return res;
}

string decode(string data){
   int digit_count = 0;
   string res;
   for(char c: data){
       if(isdigit(c)){
            if(digit_count>0)
                digit_count = 10*digit_count + (c - '0');
            else 
                digit_count = c - '0';
        } else {
            res += c;
            for(int j=0; j<digit_count-1; j++)
                res += c;
            digit_count = 0;
        }
    }
    return res;
}


}  // namespace run_length_encoding

File: solutions/free-pascal/run-length-encoding/1/RunLengthEncoding.pas
unit RunLengthEncoding;

{$mode ObjFPC}{$H+}

interface

function encode(const AString : string) : string;
function decode(const AString : string) : string;

implementation

uses SysUtils, Math;



function encode(const AString : string) : string;
var 
  ch: char; 
  count,i : integer;
begin
  result := '';
  count := 0;

  for i:=low(AString) to high(AString) do
  begin

    ch:=AString[i];
    inc(count);
    if (i=high(AString)) or (ch<>AString[i+1]) then
    begin
      if count>1 then
        result:= result + IntToStr(count);
      
      result:= result + ch;
      count:=0;
    end;
  end;
end;

function decode(const AString : string) : string;
var 
  digitCount, i : integer;
  ch : char;
begin
  result := '';
  digitCount := 0;
  for ch in AString do
  begin
    if (ord('0')<=ord(ch)) and (ord(ch)<=ord('9')) then
      begin
        if digitCount>0 then
          digitCount := 10*digitCount + (ord(ch) - ord('0'))
        else
          digitCount := ord(ch) - ord('0');
      end
    else
      begin
        for i:=0 to max(0, digitCount-1) do
        begin
          result := result + ch;
        end;
        digitCount:=0;
      end;
  end;

end;

end.
